[flaky-tests] fix ProcessGroupGlooTest

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31255 [flaky-tests] fix ProcessGroupGlooTest**

This test had 2 issues. A timeout would occasionally happen due to a timeout of 50ms, and CUDA could would get compiled and run on CPU, leading to errors. This PR fixes those issues.

Differential Revision: [D19028231](https://our.internmc.facebook.com/intern/diff/D19028231/)

rohan-varma committed:
[flaky-tests] skip ProcessGroupGlooTest if CUDA is not available

Similar to other tests, this test can error if CUDA is not avaiable.

Differential Revision: [D19028231](https://our.internmc.facebook.com/intern/diff/D19028231/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[flaky-tests] skip ProcessGroupGlooTest if CUDA is not available"


Similar to other tests, this test can error if CUDA is not avaiable.

Differential Revision: [D19028231](https://our.internmc.facebook.com/intern/diff/D19028231/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit e40f066:
* **6/6** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=e40f066dff482fc2e9a9dbb87d143c7f2885e06e).

### 4 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3949289) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (1/4)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74115629))
```
Dec 13 17:43:12 [ 95%] Built target make_cifar_db 
Dec 13 17:43:12 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/python_dimname.cpp.o 
Dec 13 17:43:12 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/QScheme.cpp.o 
Dec 13 17:43:13 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/Module.cpp.o 
Dec 13 17:43:13 Scanning dependencies of target split_db 
Dec 13 17:43:13 In file included from /var/lib/jenkins/workspace/torch/lib/c10d/test/ProcessGroupGlooTest.cpp:13: 
Dec 13 17:43:13 In file included from /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/ProcessGroupGloo.hpp:26: 
Dec 13 17:43:13 /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/Utils.hpp:225:21: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare] 
Dec 13 17:43:13   for (int i = 1; i < tensors.size(); ++i) { 
Dec 13 17:43:13                   ~ ^ ~~~~~~~~~~~~~~ 
Dec 13 17:43:13 /var/lib/jenkins/workspace/torch/lib/c10d/test/ProcessGroupGlooTest.cpp:371:18: error: no member named 'is_available' in namespace 'at::cuda' 
Dec 13 17:43:13   if (!at::cuda::is_available()) { 
Dec 13 17:43:13        ~~~~~~~~~~^ 
Dec 13 17:43:13 1 warning and 1 error generated. 
Dec 13 17:43:13 caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/build.make:62: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o' failed 
Dec 13 17:43:13 make[2]: *** [caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o] Error 1 
Dec 13 17:43:13 CMakeFiles/Makefile2:11528: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/all' failed 
Dec 13 17:43:13 make[1]: *** [caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/all] Error 2 
Dec 13 17:43:13 make[1]: *** Waiting for unfinished jobs.... 
Dec 13 17:43:13 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/PtrWrapper.cpp.o 
Dec 13 17:43:13 [ 95%] Building CXX object binaries/CMakeFiles/split_db.dir/split_db.cc.o 
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3949285) pytorch_macos_10_13_py3_build (2/4)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74116225))
```
Dec 13 09:44:42 [2828/2967] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/CudaIPCTypes.cpp.o 
Dec 13 09:44:42 [2829/2967] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/DataLoader.cpp.o 
Dec 13 09:44:43 [2830/2967] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/Device.cpp.o 
Dec 13 09:44:43 In file included from ../torch/csrc/Device.cpp:5: 
Dec 13 09:44:43 ../torch/csrc/utils/python_arg_parser.h:657:13: warning: unused function 'check_has_torch_function' [-Wunused-function] 
Dec 13 09:44:43 static auto check_has_torch_function(PyObject* obj) -> bool 
Dec 13 09:44:43             ^ 
Dec 13 09:44:43 1 warning generated. 
Dec 13 09:44:43 [2831/2967] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/Dtype.cpp.o 
Dec 13 09:44:44 [2832/2967] Building CXX object caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o 
Dec 13 09:44:44 FAILED: caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o  
extra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces "-Xpreprocessor -fopenmp -I/usr/local/include" -std=gnu++14 -MD -MT caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o -MF caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o.d -o caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o -c ../torch/lib/c10d/test/ProcessGroupGlooTest.cpp 
Dec 13 09:44:44 In file included from ../torch/lib/c10d/test/ProcessGroupGlooTest.cpp:13: 
Dec 13 09:44:44 In file included from ../torch/lib/c10d/../c10d/ProcessGroupGloo.hpp:26: 
Dec 13 09:44:44 ../torch/lib/c10d/../c10d/Utils.hpp:225:21: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare] 
Dec 13 09:44:44   for (int i = 1; i < tensors.size(); ++i) { 
Dec 13 09:44:44                   ~ ^ ~~~~~~~~~~~~~~ 
Dec 13 09:44:44 ../torch/lib/c10d/test/ProcessGroupGlooTest.cpp:371:18: error: no member named 'is_available' in namespace 'at::cuda' 
Dec 13 09:44:44   if (!at::cuda::is_available()) { 
Dec 13 09:44:44        ~~~~~~~~~~^ 
Dec 13 09:44:44 1 warning and 1 error generated. 
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3949301) pytorch_xla_linux_xenial_py3_6_clang7_build (3/4)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74116895))
```
Dec 13 17:45:05 static auto check_has_torch_function(PyObject* obj) -> bool 
Dec 13 17:45:05             ^ 
Dec 13 17:45:05 1 warning generated. 
Dec 13 17:45:05 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/python_dimname.cpp.o 
Dec 13 17:45:05 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/QScheme.cpp.o 
Dec 13 17:45:06 In file included from /var/lib/jenkins/workspace/torch/lib/c10d/test/ProcessGroupGlooTest.cpp:13: 
Dec 13 17:45:06 In file included from /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/ProcessGroupGloo.hpp:26: 
Dec 13 17:45:06 /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/Utils.hpp:225:21: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare] 
Dec 13 17:45:06   for (int i = 1; i < tensors.size(); ++i) { 
Dec 13 17:45:06                   ~ ^ ~~~~~~~~~~~~~~ 
Dec 13 17:45:06 /var/lib/jenkins/workspace/torch/lib/c10d/test/ProcessGroupGlooTest.cpp:371:18: error: no member named 'is_available' in namespace 'at::cuda' 
Dec 13 17:45:06   if (!at::cuda::is_available()) { 
Dec 13 17:45:06        ~~~~~~~~~~^ 
Dec 13 17:45:06 1 warning and 1 error generated. 
Dec 13 17:45:06 caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/build.make:62: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o' failed 
Dec 13 17:45:06 make[2]: *** [caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o] Error 1 
Dec 13 17:45:06 make[1]: *** [caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/all] Error 2 
Dec 13 17:45:06 make[1]: *** Waiting for unfinished jobs.... 
Dec 13 17:45:06 CMakeFiles/Makefile2:11505: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/all' failed 
Dec 13 17:45:06 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/Module.cpp.o 
Dec 13 17:45:06 [ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/PtrWrapper.cpp.o 
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3949296) pytorch_linux_xenial_py2_7_9_build (4/4)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74117061))
```
Dec 13 17:45:09 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/python_dimname.cpp.o 
Dec 13 17:45:09 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/QScheme.cpp.o 
Dec 13 17:45:09 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/Module.cpp.o 
Dec 13 17:45:10 [ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/PtrWrapper.cpp.o 
Dec 13 17:45:10 [ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/Size.cpp.o 
Dec 13 17:45:10 [ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/Storage.cpp.o 
Dec 13 17:45:10 [ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/api/src/python/init.cpp.o 
Dec 13 17:45:11 [ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/autograd/functions/init.cpp.o 
Dec 13 17:45:11 [ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/autograd/init.cpp.o 
Dec 13 17:45:11 /var/lib/jenkins/workspace/torch/lib/c10d/test/ProcessGroupGlooTest.cpp: In function 'int main(int, char**)': 
Dec 13 17:45:11 /var/lib/jenkins/workspace/torch/lib/c10d/test/ProcessGroupGlooTest.cpp:371:18: error: 'is_available' is not a member of 'at::cuda' 
Dec 13 17:45:11    if (!at::cuda::is_available()) { 
Dec 13 17:45:11                   ^~~~~~~~~~~~ 
Dec 13 17:45:11 make[2]: *** [caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o] Error 1 
Dec 13 17:45:11 caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/build.make:62: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o' failed 
Dec 13 17:45:11 make[1]: *** [caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/all] Error 2 
Dec 13 17:45:11 CMakeFiles/Makefile2:11546: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/all' failed 
Dec 13 17:45:11 make[1]: *** Waiting for unfinished jobs.... 
Dec 13 17:45:11 [ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/autograd/python_anomaly_mode.cpp.o 
Dec 13 17:45:12 [ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/autograd/python_cpp_function.cpp.o 
Dec 13 17:45:12 [ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/autograd/python_engine.cpp.o 
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

rohan-varma committed:
Update on "[flaky-tests] skip ProcessGroupGlooTest if CUDA is not available"


Similar to other tests, this test can error if CUDA is not avaiable.

Differential Revision: [D19028231](https://our.internmc.facebook.com/intern/diff/D19028231/)

[ghstack-poisoned]

rohan-varma commented:
Hmm, this is strange. The test failures are legit but I'm using the same call as in https://github.com/pytorch/pytorch/pull/30345 which appears to be valid...

rohan-varma committed:
Update on "[flaky-tests] skip ProcessGroupGlooTest if CUDA is not available"


Similar to other tests, this test can error if CUDA is not avaiable.

Differential Revision: [D19028231](https://our.internmc.facebook.com/intern/diff/D19028231/)

[ghstack-poisoned]

pritamdamania87 commented:
> Hmm, this is strange. The test failures are legit but I'm using the same call as in #30345 which appears to be valid...

Hmm might be something to do for builds which don't have CUDA enabled, can you try torch::cuda::is_available?

pietern commented:
@rohan-varma Why can it fail? It should be a CPU only test.

pritamdamania87 commented:
Can we figure out a way to fix the CUDA tests here and land the PR? According to our test dashboard, this is the only test that is flaky internally :)

rohan-varma committed:
Update on "[flaky-tests] skip ProcessGroupGlooTest if CUDA is not available"


Similar to other tests, this test can error if CUDA is not avaiable.

Differential Revision: [D19028231](https://our.internmc.facebook.com/intern/diff/D19028231/)

[ghstack-poisoned]

rohan-varma commented:
@pietern I'm not sure, it appears that the test is CPU-only. Though if you look at the internal test failure, it appears to be the same CUDA issue we ran into in other tests (and resolved via PRs such as https://github.com/pytorch/pytorch/pull/30345/files). If this PR is patched, the tests stop failing.

rohan-varma commented:
Actually, this isn't the right approach to skip the test. If we do this, the test will only run on GPU devices but we also want to run it on CPU (as @pietern mentioned). 

The issue in the test is related to the  `#ifdef USE_CUDA` and the following `testAllreduce(file.path, at::DeviceType::CUDA);`, which seems to be getting compiled even on CPU.

rohan-varma committed:
Update on "[flaky-tests] skip ProcessGroupGlooTest if CUDA is not available"


Similar to other tests, this test can error if CUDA is not avaiable.

Differential Revision: [D19028231](https://our.internmc.facebook.com/intern/diff/D19028231/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -39,7 +41,7 @@ class SignalTest {
     // Use tiny timeout to make this test run fast
     ::c10d::ProcessGroupGloo::Options options;
-    options.timeout = std::chrono::milliseconds(50);
+    options.timeout = std::chrono::milliseconds(1000);
```
nit: Add a comment here explaining the timeout.

pritamdamania87 reviewed (approved):
Looks good and all tests are passing!

rohan-varma committed:
Update on "[flaky-tests] fix ProcessGroupGlooTest"


This test had 2 issues. A timeout would occasionally happen due to a timeout of 50ms, and CUDA could would get compiled and run on CPU, leading to errors. This PR fixes those issues.

Differential Revision: [D19028231](https://our.internmc.facebook.com/intern/diff/D19028231/)

[ghstack-poisoned]
